src/Rewards.sol

```diff
@@ -0,0 +1,239 @@
+
+    /**
+     * @dev Error when the recipient's reward sequence does not match.
+     */
+    error InvalidRecipientSequence();
+
+    /**
+     * @dev Event emitted when the reward quota is set.
+     */
+    event RewardQuotaSet(uint256 quota);
+
+    /**
+     * @dev Event emitted when a reward is minted.
+     */
+    event RewardMinted(address indexed recipient, uint256 amount, uint256 totalRewardsClaimed);
+
+    /**
+     * @dev Initializes the contract with the specified parameters.
+     * @param token Address of the NODL token contract.
+     * @param initialQuota Initial reward quota.
+     * @param initialPeriod Initial reward period.
+     * @param oracleAddress Address of the authorized oracle.
+     */
+    constructor(NODL token, uint256 initialQuota, uint256 initialPeriod, address oracleAddress)
+        EIP712(SIGNING_DOMAIN, SIGNATURE_VERSION)
+    {
+        // This is to avoid the ongoinb overhead of safe math operations
+        if (initialPeriod == 0) {
+            revert ZeroPeriod();
+        }
+        // This is to prevent overflows while avoiding the ongoing overhead of safe math operations
+        if (initialPeriod > MAX_PERIOD) {
+            revert TooLongPeriod();
+        }
+
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+
+        nodlToken = token;
+        rewardQuota = initialQuota;
+        rewardPeriod = initialPeriod;
+        rewardsClaimed = 0;
+        quotaRenewalTimestamp = block.timestamp + rewardPeriod;
+        authorizedOracle = oracleAddress;
+    }
+
+    /**
+     * @dev Sets the reward quota. Only accounts with the QUOTA_SETTER_ROLE can call this function.
+     * @param newQuota The new reward quota.
+     */
+    function setRewardQuota(uint256 newQuota) external {
+        _checkRole(DEFAULT_ADMIN_ROLE);
+        rewardQuota = newQuota;
+        emit RewardQuotaSet(newQuota);
+    }
+
+    /**
+     * @dev Mints rewards to the recipient if the signature is valid and quota is not exceeded.
+     * @param reward The reward details.
+     * @param signature The signature from the authorized oracle.
+     */
+    function mintReward(Reward memory reward, bytes memory signature) external {
+        _mustBeFromAuthorizedOracle(digest(reward), signature);
+
+        _mustBeExpectedSequence(reward.recipient, reward.sequence);
+
+        if (block.timestamp >= quotaRenewalTimestamp) {
+            rewardsClaimed = 0;
+
+            // The following operations are safe based on the constructor's requirements for longer than the age of universe :)
+            uint256 timeAhead = block.timestamp - quotaRenewalTimestamp;
+            quotaRenewalTimestamp = block.timestamp + rewardPeriod - (timeAhead % rewardPeriod);
+        }
+
+        (bool sucess, uint256 newRewardsClaimed) = rewardsClaimed.tryAdd(reward.amount);
+        if (!sucess || newRewardsClaimed > rewardQuota) {
+            revert RewardQuotaExceeded();
+        }
+        rewardsClaimed = newRewardsClaimed;
```

Comment:
Not a blocker, but I think you could avoid the new variable allocation below via. To be checked as I have not ran this myself.
```suggestion
        (bool sucess, rewardsClaimed) = rewardsClaimed.tryAdd(reward.amount);
        if (!sucess || rewardsClaimed > rewardQuota) {
            revert RewardQuotaExceeded();
        }
```

Reply:
This would not work I had tried it before.

Reply:
Understood